src/field/battle_tower.c

```diff
@@ -0,0 +1,2375 @@
+	}
+
+	SetBattleTowerRecordChecksum(&gSaveBlock2.battleTower.playerRecord);
+	sub_8135A3C();
+}
+
+void sub_8135BA0(void)
+{
+	u8 battleTowerLevelType = gSaveBlock2.battleTower.battleTowerLevelType;
+
+	if (gSpecialVar_0x8004 == 3 || gSpecialVar_0x8004 == 0)
+	{
+		if (gSaveBlock2.battleTower.curStreakChallengesNum[battleTowerLevelType] > 1
+			|| gSaveBlock2.battleTower.curChallengeBattleNum[battleTowerLevelType] > 1)
+		{
+			sub_8135AC4();
+		}
+	}
+
+	sub_8135CFC();
+
+	gSaveBlock2.battleTower.battleOutcome = gBattleOutcome;
+
+	if (gSpecialVar_0x8004 != 3)
+	{
+		gSaveBlock2.battleTower.var_4AE[battleTowerLevelType] = gSpecialVar_0x8004;
+	}
+
+	VarSet(VAR_0x4000, 0);
+	gSaveBlock2.battleTower.unk_554 = 1;
+	TrySavingData(EREADER_SAVE);
+}
+
+void BattleTower_SoftReset(void)
+{
+	DoSoftReset();
+}
+
+void ValidateBattleTowerRecordChecksums(void)
+{
+	u32 i;
+	s32 recordIndex;
+	struct BattleTowerRecord *record;
+	u32 checksum;
+
+	checksum = 0;
+	for (i = 0; i < (sizeof(struct BattleTowerRecord) / sizeof(u32)) - 1; i++)
+	{
+		checksum += ((u32 *)&gSaveBlock2.battleTower.playerRecord)[i];
+	}
+
+	if (gSaveBlock2.battleTower.playerRecord.checksum != checksum)
+	{
+		ClearBattleTowerRecord(&gSaveBlock2.battleTower.playerRecord);
+	}
+
+	for (recordIndex = 0; recordIndex < 5; recordIndex++)
+	{
+		record = &gSaveBlock2.battleTower.records[recordIndex];
+		checksum = 0;
+		for (i = 0; i < (sizeof(struct BattleTowerRecord) / sizeof(u32)) - 1; i++)
+		{
+			checksum += ((u32 *)record)[i];
+		}
+
+		if (gSaveBlock2.battleTower.records[recordIndex].checksum != checksum)
+		{
+			ClearBattleTowerRecord(&gSaveBlock2.battleTower.records[recordIndex]);
+		}
+	}
+}
+
+void SetBattleTowerRecordChecksum(struct BattleTowerRecord *record)
+{
+	u32 i;
+
+	record->checksum = 0;
+	for (i = 0; i < (sizeof(struct BattleTowerRecord) / sizeof(u32)) - 1; i++)
+	{
+		record->checksum += ((u32 *)record)[i];
+	}
+}
+
+void ClearBattleTowerRecord(struct BattleTowerRecord *record)
+{
+	u32 i;
+
+	for (i = 0; i < sizeof(struct BattleTowerRecord) / sizeof(u32); i++)
+	{
+		((u32 *)record)[i] = 0;
+	}
+}
+
+void sub_8135CFC(void)
+{
+	s32 i;
+
+	get_trainer_name(gSaveBlock2.battleTower.defeatedByTrainerName);
+	gSaveBlock2.battleTower.defeatedBySpecies = gBattleMons[1].species;
+	gSaveBlock2.battleTower.firstMonSpecies = gBattleMons[0].species;
+
+	for (i = 0; i < POKEMON_NAME_LENGTH; i++)
+	{
+		gSaveBlock2.battleTower.firstMonNickname[i] = gBattleMons[0].nickname[i];
+	}
+}
+
+u16 GetCurrentBattleTowerWinStreak(u8 battleTowerLevelType)
+{
+	u16 winStreak = ((gSaveBlock2.battleTower.curStreakChallengesNum[battleTowerLevelType] - 1) * 7 - 1)
+					+ gSaveBlock2.battleTower.curChallengeBattleNum[battleTowerLevelType];
+
+	if (winStreak > 9999)
+	{
+		return 9999;
+	}
+
+	return winStreak;
+}
+
+#ifdef NONMATCHING
+void DetermineBattleTowerPrize(void)
+{
+    u16 prizeItem;
+    struct SaveBlock2 *saveBlock = &gSaveBlock2;
```

Comment:
Game Freak has never, ever, done a legitimate gSaveBlock recast to a struct pointer. This is most likely not what they did. You have to use it directly somehow.

Reply:
I agree, but I've exhausted all my ideas.  Currently, this function just swap the `bl Random()` with the `ldr r4, _08135DE8 @ =sShortStreakPrizes` lines.  Otherwise, identical.